spec/ruby_speech/grxml/grammar_spec.rb

```diff
@@ -232,6 +232,111 @@ module GRXML
           grammar.inline!.should == inline_grammar
           grammar.should == inline_grammar
         end
+
+        context "nested" do
+          context "in a self-referencial infinite loop" do
+            subject do
+              RubySpeech::GRXML.draw mode: :dtmf, root: "main" do
+                rule id: :main, scope: "public" do
+                  ruleref uri: "#paradox"
+                end
+                rule id: "paradox" do
+                  ruleref uri: "#paradox"
+                end
+              end
+            end
+
+            it "should be rejected with an error" do
+              expect { subject.inline }.to raise_error ArgumentError, /Max ruleref recursion level of 25 has been exceeded./
+            end
+          end
+
+          context "with an invalid-reference" do
+            subject do
+              RubySpeech::GRXML.draw mode: :dtmf, root: "main" do
+                rule id: :main, scope: "public" do
+                  ruleref uri: "#lost"
+                end
+              end.inline
+            end
+
+            it "should raise a descriptive exception" do
+              expect { subject }.to raise_error ArgumentError, 'The Ruleref "#lost" is referenced but not defined'
+            end
+          end
+
+          context "deeply" do
+            before :each do
+              subject.root = "main"
+              subject << Rule.new(doc, id: "main", scope: "public") do
+                ruleref uri: "#level0"
+              end
+
+              levels.times do |i|
+                next_ref = i + 1
+                subject << if next_ref < levels
+                   Rule.new(doc, id: "level#{i}") do
+                    ruleref uri: "#level#{next_ref}"
+                  end
```

Comment:
`end` at 280, 18 is not aligned with `Rule.new(doc, id: "level#{i}") do` at 278, 19.